### PR TITLE
MY9231/MY9291 LED driver support

### PR DESCRIPTION
**Pull request by puuu**
## Description:
Implementation of  MY-SEMI MY9231 family LED driver.
Should support at least the following driver modules:
- MY9291 (4 channels)
- MY9231 (3 channels)
    
They are used in e.g., the following light blub:
- Sonoff B1 (MY9231)
- Ai-Thinker AiLight WiFi light bulb (MY9291)
- Arilux E27 Smart Bulb (MY9231)

Tested with Sonoff B1 r2.
Code is adapted from [MY92XX LED driver library for Arduino AVR and ESP8266](https://github.com/xoseperez/my92xx). Unfortunately, there is no free detailed Datasheet available and the descriptions of the [MY9221](https://raw.githubusercontent.com/SeeedDocument/Grove-LED_Bar/master/res/MY9221_DS_1.0.pdf) seems not be compatible with the MY9231.

**Related issue (if applicable):** fixes

**Pull request in [esphomedocs](https://github.com/OttoWinter/esphomedocs) with documentation (if applicable):** OttoWinter/esphomedocs#80
**Pull request in [esphomeyaml](https://github.com/OttoWinter/esphomeyaml) with YAML changes (if applicable):** OttoWinter/esphomeyaml#227

## Example entry for YAML configuration (if applicable):
```yaml
my9231:
  data_pin: GPIO12
  clock_pin: GPIO14
  num_channels: 6
  num_chips: 2

output:
  - platform: my9231
    id: output_blue
    channel: 0
  - platform: my9231
    id: output_red
    channel: 1
  - platform: my9231
    id: output_green
    channel: 2
  - platform: my9231
    id: output_warm_white
    channel: 4
  - platform: my9231
    id: output_cold_white
    channel: 5

light:
  - platform: rgbww
    name: "Sonoff B1"
    red: output_red
    green: output_green
    blue: output_blue
    cold_white: output_cold_white
    warm_white: output_warm_white
    cold_white_color_temperature: 6500 K
    warm_white_color_temperature: 2800 K
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Check this box if you have read, understand, comply, and agree with the [Code of Conduct](https://github.com/OttoWinter/esphomelib/blob/master/CODE_OF_CONDUCT.md).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphomedocs](https://github.com/OttoWinter/esphomedocs).

### CODE REVIEW

File: src/esphomelib/output/my9231_output_component.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  };
+
+ protected:
+  void set_channel_value(uint8_t channel, uint16_t value);
+  void init_chips(uint8_t cmd);
+  void write_word(uint16_t value, uint8_t bits);
+  void send_di_pulses(uint8_t count);
+
+  GPIOOutputPin pin_di_;
+  GPIOOutputPin pin_dcki_;
```
Comment (OttoWinter) [inline src/esphomelib/output/my9231_output_component.h]:
```cpp
  GPIOPin *pin_di_;
  GPIOPin *pin_dcki_;
```

File: src/esphomelib/output/my9231_output_component.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+   * @param pin_di The pin which DI is connected to.
+   * @param pin_dcki The pin which DCKI is connected to.
+   * @param num_channels Total number of channels of the whole daisy chain.
+   * @param num_chips Number of chips in the daisy chain.
+   * @param bit_depth Bit depth of each channel
+   * @param update Update/reset duty data at boot (driver will keep
+   *               configuration after powercycle)
+   */
+  MY9231OutputComponent(const GPIOOutputPin &pin_di,
+                        const GPIOOutputPin &pin_dcki,
```
Comment (OttoWinter) [inline src/esphomelib/output/my9231_output_component.h]:
```cpp
  MY9231OutputComponent(GPIOPin *pin_di, GPIOPin *pin_dcki,
```

File: src/esphomelib/application.cpp [hunk 3]
```diff
@@ -240,6 +240,14 @@ PCA9685OutputComponent *Application::make_pca9685_component(float frequency) {
 }
 #endif
 
+#ifdef USE_MY9231_OUTPUT
+MY9231OutputComponent *Application::make_my9231_component(const GPIOOutputPin &pin_di,
+                                                          const GPIOOutputPin &pin_dcki) {
+  auto *my9231 = new MY9231OutputComponent(pin_di, pin_dcki);
```
Comment (OttoWinter) [inline src/esphomelib/application.cpp]:
```suggestion
  auto *my9231 = new MY9231OutputComponent(pin_di.copy(), pin_dcki.copy());
```

File: src/esphomelib/output/my9231_output_component.cpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    this->pin_di_.digital_write(value & (1 << (i - 1)));
+    this->pin_dcki_.digital_write(!this->pin_dcki_.digital_read());
+  }
+}
+
+void MY9231OutputComponent::send_di_pulses(uint8_t count) {
+  delayMicroseconds(12);
+  for (uint8_t i = 0; i < count; i++) {
+    this->pin_di_.digital_write(true);
+    this->pin_di_.digital_write(false);
```
Comment (OttoWinter) [inline src/esphomelib/output/my9231_output_component.cpp]: What is the timing requirement of these pulses? In theory (if all optimizations are applied), this loop could run at around 20-40MHz.

File: src/esphomelib/output/my9231_output_component.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    : pin_di_(pin_di),
+      pin_dcki_(pin_dcki),
+      num_channels_(num_channels),
+      num_chips_(num_chips),
+      bit_depth_(bit_depth),
+      update_(update) {}
+
+void MY9231OutputComponent::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up MY9231OutputComponent.");
+  this->pin_di_.setup();
```
Comment (OttoWinter) [inline src/esphomelib/output/my9231_output_component.cpp]:
```suggestion
  this->pin_di_->setup();
```

Same for other calls.

**Review by OttoWinter — COMMENTED** [review]
Looks almost perfect 😺

Just one thing regarding the use of `GPIOOutputPin`. If you store it by value as you do in `my9231_output_component.h`, this component will only be usable from GPIO pins on the ESP, not IO expanders. Therefore esphomelib uses points which store the type for virtual calls (exception is esp8266_pwm, where the only pins that are supported *are* internal GPIO pins)

File: src/esphomelib/output/my9231_output_component.cpp (hunk 4 above)
Comment (OttoWinter) [reply src/esphomelib/output/my9231_output_component.cpp]: Ok, let's just leave it as is then 👍

File: src/esphomelib/output/my9231_output_component.cpp [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+float MY9231OutputComponent::get_setup_priority() const {
+  return setup_priority::HARDWARE;
+}
+
+void MY9231OutputComponent::set_channel_value(uint8_t channel, uint16_t value) {
+  ESP_LOGV(TAG, "set channels %u to %u", channel, value);
+  if (this->pwm_amounts_[channel] != value) {
+    this->update_ = true;
+  }
+  this->pwm_amounts_[this->num_channels_ - channel - 1] = value;
```
[the remark this answers is not in the document]
Comment (puuu) [reply src/esphomelib/output/my9231_output_component.cpp]: Good catch! Thank you! I overlooked that when I changed the indexing :-(.
Fixed.

**Review by OttoWinter — APPROVED** [review]
Ok, LGTM 👍

I will merge this now, so that it can be part of 1.9.0 which I hope will be ready in this week (depends on the amount of bugs people find of course).

I've made some pretty big refactorings recently (https://github.com/OttoWinter/esphomelib/pull/267, https://github.com/OttoWinter/esphomeyaml/pull/229), so I will merge this as-is and then make small update commits to refactor to the new system 👍

Thanks a lot for this new integration!

**Comment by puuu** [comment]
FYI, I have checked the last changes in `dev` and it works fine.

Thank you for sharing this amazing project. It was a pleasure to support it.

Since I saw all the new features documented in the `next` docs, I am looking forward for the next official version.

**Pull request merged**